src/engine/threaded_engine.h

```diff
@@ -60,6 +60,9 @@ namespace engine {
 // Forward declarations
 struct ThreadedOpr;
 
+/*! shared_ptr to exception_ptr, used for exception handling */
+typedef std::shared_ptr<std::exception_ptr> ExceptionRef;
```

Comment:
Do we need to wrap it in a shared_ptr? exception_ptr has already shared ptr semantics according to cppreference.

Reply:
Also, the name Ref to me is confusing, why not Ptr? why add a suffix of the type at all?

Reply:
exception_ptr cannot be dereferenced , so we cannot update the exception object it is pointing to or make it nullptr. Since this is a requirement for us we wrapped it in a shared_ptr. Used ref to make it consistent with other places in MXNet.